doc/roadmap-2024.md

```diff
@@ -0,0 +1,75 @@
+# TiKV Roadmap (2024~2025)
+
+- Last Updated: 2024-04-23
+
+This is a roadmap of TiKV project from engineering and architecture aspects. It expresses the intension of how we will improve this project to make TiKV fit more users' use cases in next 1 to 2 years. The roadmap includes several domains like latency and performance improvement, stability, observability, etc. Notice: all content list in the roadmap is our intension, we will try our best efforts to make it happen, but it doesn't mean we have a commitment for it, please don't rely on features/capabilities currently TiCDC doesn't have for your projects which have strick timeline requirement, even these features listed in the roadmap.
+
+## QoS: Single-Digit Millsecond Predictable Latency
+
+This domain aims to enhance TiKV to provide predictable single-digit millisecond latency under different scenarios like too many MVCC versions, disk IO temporarily jitter, high read QPS in a small data range, etc. 
+
+### In-Memory Engine
+
+TiKV tends to pileup many MVCC versions for a data range if there are high frequency UPDATE operations in this specified data range. It will slowdown the scan/read speed dramatically in this data range, and easily cause high read latency which may cause applications failure/timeout. The In-Memory Engine was designed to accelerate the scan/read speed when there are too many MVCC versions. The In-Memory Engine only cache latest few versions for each row which can reduce the amount of retrieving KVs for most read/scan queries.
+
+### Mitigate the Impact of Temporary Disk IO Jitter
+
+IO latency is very critical to the QoS of TiKV, because TiKV highly depends on the persistence of raftlog and read data from disk(in case of data not in block-cache). If there is a disk IO jitter, the application may suffer from high latency of queries. As the total number of TiKV nodes increases for a TiKV cluster, the higher possibility the TiKV cluster will encouter disk IO jitter issues. We experienced more disk IO jitter issues in cloud disk like AWS EBS compare to local nvm-e disk. We will improve the capability of tolerating single Disk IO jitter issues in a cluster by introduce improvements like apply-before-persistent, mitigate-latency-amplification, etc.
+
+### Low Read Latency of Hot Region
+
+The read queries evenly distributed in different data ranges is a ideally workload, it can make good use of all resources across different TiKV nodes. But in the real world, queries tend to skew to some small data ranges, like latest added/modified data has higher possibility to be retrieved, a specified category's data has higher read queries in some special time range, etc. We call these highly retrieved data ranges as hot regions. Hot regions tend to cause higher query latency since there might be some queries need to wait. A faster in-memory engine cache layer might have a great benefit in this hot region case since it will use less resource to serve a query.
```

Comment:
```suggestion
Read queries being evenly distributed in different data ranges is an ideal workload, which can make good use of all resources across different TiKV nodes. But in the real world, queries tend to skew to some small data ranges, like latest added/modified data has higher possibility to be retrieved, a specified category's data has higher read queries in some special time range, etc. We call these highly retrieved data ranges as hot regions. Hot regions tend to cause higher query latency since there might be some queries need to wait. A faster in-memory engine cache layer might have a great benefit in this hot region case since it will use less resource to serve a query.
```